js/src/hooks/useCroppedImageSelector/useCroppedImageSelector.js

```diff
@@ -0,0 +1,289 @@
+ *
+ * @param {number} width The actual width.
+ * @param {number} height The actual height.
+ * @param {number} expectedWidth The expected width.
+ * @param {number} expectedHeight The expected height.
+ * @return {number} The percent error of aspect ratio.
+ */
+export function calcRatioPercentError(
+	width,
+	height,
+	expectedWidth,
+	expectedHeight
+) {
+	const imageRatio = width / height;
+	const expectedRatio = expectedWidth / expectedHeight;
+	const errorRatio =
+		imageRatio > expectedRatio
+			? imageRatio / expectedRatio
+			: expectedRatio / imageRatio;
+
+	return ( errorRatio - 1 ) * 100;
+}
+
+/**
+ * Hook for opening a modal to select images with fixed aspect ratio via
+ * WordPress Media library. It will request cropping when the size of the
+ * selected image doesn't match the given aspect ratio.
+ *
+ * @param {Object} options
+ * @param {number} options.minWidth The minimum width and also the width to be calculated as the aspect ratio.
+ * @param {number} options.minHeight The minimum height and also the height to be calculated as the aspect ratio.
+ * @param {number} options.suggestedWidth The suggested width showing on the modal.
+ * @param {number} options.suggestedHeight The suggested height showing on the modal.
+ * @param {(image: ImageMedia) => void} options.onSelect Callback when an image is selected.
+ * @param {(image: ImageMedia) => void} options.onDelete Callback when an image is deleted.
+ * @param {number} [options.allowedRatioPercentError=1] The percent error of the aspect ratio.
+ * @return {CroppedImageSelector} The handle of this hook.
+ */
+export default function useCroppedImageSelector( {
+	minWidth,
+	minHeight,
+	suggestedWidth,
+	suggestedHeight,
+	onSelect,
+	onDelete,
+	allowedRatioPercentError = 1,
+} ) {
+	const callbackRef = useRef( {} );
+	callbackRef.current.onSelect = onSelect;
+	callbackRef.current.onDelete = onDelete;
+
+	const openSelector = useCallback(
+		( preselectedId ) => {
+			const { media } = wp;
+			const sizeErrorMessage = sprintf(
+				// translators: 1: Minimum width, 2: Minimum height.
+				__(
+					'Image size needs to be at least %1$d x %2$d',
+					'google-listings-and-ads'
+				),
+				minWidth,
+				minHeight
+			);
+
+			// Will be called by the controller of the parent class of CustomizeImageCropper. Ref:
+			// - https://github.com/WordPress/wordpress-develop/blob/5.9.0/src/js/media/controllers/customize-image-cropper.js#L14
+			// - https://github.com/WordPress/wordpress-develop/blob/5.9.0/src/js/media/controllers/cropper.js#L65
+			// - https://github.com/WordPress/wordpress-develop/blob/5.9.0/src/js/media/views/cropper.js#L48-L54
+			const imgSelectOptions = ( attachment, controller ) => {
+				const width = attachment.get( 'width' );
+				const height = attachment.get( 'height' );
+				const args = [ width, height, minWidth, minHeight ];
+				const ratioPercentError = calcRatioPercentError( ...args );
+
+				if ( ratioPercentError < allowedRatioPercentError ) {
```

Comment:
💅  I guess this should be `<= `since we "allowed" that percent error, so equals is valid. Otherwise we could rename it to `ratioPercentErrorTolerance` or just `ratioPercentError`

Reply:
Thanks for the suggestion. Renamed in https://github.com/woocommerce/google-listings-and-ads/pull/1819/commits/4f321138ba65ec29d6246d4f68d9352edde72702.